### ISSUE TITLE
"local variable 'result' referenced before assignment" compiler error with auto_pickle in 0.26

### ISSUE THREAD

**Issue by jamadden** [cython/cython#1786]
gevent [fails to compile](https://travis-ci.org/gevent/gevent/jobs/255646182#L538) with cython 0.26 (observed on OS X, Windows and Linux): 
```
    cython -o gevent.ares.c src/gevent/ares.pyx
    
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
    def __pyx_unpickle_result(__pyx_type, long __pyx_checksum, __pyx_state):
        if __pyx_checksum != 0xc9b5cdc:
            from pickle import PickleError
            raise PickleError("Incompatible checksums (%s vs 0xc9b5cdc = (exception, value))" % __pyx_checksum)
        result = result.__new__(__pyx_type)
                      ^
    ------------------------------------------------------------
    
    (tree fragment):5:19: local variable 'result' referenced before assignment
    
    Error compiling Cython file:
    ------------------------------------------------------------
    ...
        if __pyx_checksum != 0xc9b5cdc:
            from pickle import PickleError
            raise PickleError("Incompatible checksums (%s vs 0xc9b5cdc = (exception, value))" % __pyx_checksum)
        result = result.__new__(__pyx_type)
        if __pyx_state is not None:
            __pyx_unpickle_result__set_state(<result> result, __pyx_state)
```

Where `result` is a cdef class. It seems like the name `result` itself is special. This produces it:
```
$ cat /tmp/ares.pyx
cdef class result:
	pass
$ cython -o /tmp/a.c /tmp/ares.pyx

Error compiling Cython file:
------------------------------------------------------------
...
def __pyx_unpickle_result(__pyx_type, long __pyx_checksum, __pyx_state):
    if __pyx_checksum != 0xd41d8cd:
        from pickle import PickleError
        raise PickleError("Incompatible checksums (%s vs 0xd41d8cd = ())" % __pyx_checksum)
    result = result.__new__(__pyx_type)
                  ^
------------------------------------------------------------

(tree fragment):5:19: local variable 'result' referenced before assignment

Error compiling Cython file:
------------------------------------------------------------
...
    if __pyx_checksum != 0xd41d8cd:
        from pickle import PickleError
        raise PickleError("Incompatible checksums (%s vs 0xd41d8cd = ())" % __pyx_checksum)
    result = result.__new__(__pyx_type)
    if __pyx_state is not None:
        __pyx_unpickle_result__set_state(<result> result, __pyx_state)
                                         ^
------------------------------------------------------------

(tree fragment):7:42: 'result' is not a type identifier
```

If I set `auto_pickle` to False in the directives, the class complies fine. Likewise, if I use a different name for the class ("result2"), it compiles fine:
```
$ cat /tmp/ares2.pyx
cdef class result2:
	pass
$ cython -o /tmp/a.c /tmp/ares2.pyx
$ 
```

**Comment by orivej** [comment]
This also breaks gevent: https://github.com/gevent/gevent/blob/5a6bdd676493259ea6c59a846aeaa574f744ab44/src/gevent/ares.pyx#L169

**Comment by scoder** [comment]
This looks like normal Python semantics. If you assign to a variable, it becomes a local variable. The fix is to rename the local variable so that it does not shadow the global type name any more. Don't know what changed in Cython 0.26 that prevented it from showing up before, but the bug is definitely not in Cython.

**Comment by jamadden** [comment]
The code in question is *not* part of the application. It's (incorrectly) auto-generated by Cython. There's a clash in the names generated by Cython. 

Cython 0.26 will fail to compile this trivial file:
```python
cdef class result:
   pass
```

Adding `#cython: auto-pickle=False` to the top of the file resolves the issue (because the incorrect code is no longer generated). Likewise, changing the name of the trivial class to `result2` also works (because the name no longer clashes).

Having a class named `result` shouldn't break Cython, IMHO.

**Comment by scoder** [comment]
Sorry, yes. You were very explicit about what the problem was and I didn't read well enough. Thanks for the report and the fix in #1789.
